### PR TITLE
cache block cids for faster everything

### PR DESCRIPTION
**Pull request by whyrusleeping**
improves chain validation part of sync time by a lot.

### CODE REVIEW

**Comment by whyrusleeping** [comment]
The caveat here is that we will need to ensure that once a block is created (and its cid is cached) we never change it. This should probably be done by turning block into a protected type (private fields, controlled creation, getters and such).

For now, i've included the 'paranoid' flag that degraded performance, but will panic if someone changed a block after it was cached.

File: types/block.go [hunk 1]
```diff
@@ -54,13 +54,47 @@ type Block struct {
...
 	Proof proofs.PoStProof `json:"proof"`
+
+	cachedCid cid.Cid
+
+	cachedBytes []byte
 }
 
+// set this to true to panic if the blocks data differs from the cached cid. This should
+// be obviated by changing the block to have protected construction, private fields, and
+// getters for all the values.
```
Comment (phritz) [inline types/block.go]: can you file an issue for this against testnet and link to it here? hidden technical debt is what sunk the titanic.

File: types/block.go [hunk 2]
```diff
@@ -54,13 +54,47 @@ type Block struct {
...
+const paranoid = false
```
Comment (phritz) [inline types/block.go]: please add a test that uses assert.panics to verify this behavior

**Comment by whyrusleeping** [comment]
woo! green tests!

**Pull request merged**